sarif/src/main/java/com/jetbrains/qodana/sarif/baseline/Baseline.kt

```diff
@@ -54,41 +40,41 @@ internal class DiffState(private val options: Options) {
 
 /** CAUTION: This mutates results in report and baseline **/
 internal fun applyBaseline(report: Run, baseline: Run, options: Options): DiffState {
+    val state = DiffState(options)
+
     val reportDescriptors = DescriptorLookup(report)
     val baselineDescriptors = DescriptorLookup(baseline)
     val reportIndex = FingerprintIndex()
     val baselineCounter = Counter<ResultKey>()
 
-    // shallow copies, to not mess with the underlying reports
     val undecidedFromReport = report.results.noNulls()
         .filterNot { it.baselineState == BaselineState.ABSENT }
         .onEach { result ->
-            result.equalIndicators.forEach { print -> reportIndex.add(print, result) }
+            result.equalIndicators
+                .forEach { print -> reportIndex.add(print, result) }
         }
-        .toMutableList()
+        .toCollection(IdentitySet(report.results?.size ?: 0))
 
     val undecidedFromBaseline = baseline.results.noNulls()
         .filterNot { it.baselineState == BaselineState.ABSENT }
         .onEach { result -> baselineCounter.increment(ResultKey(result)) }
-        .toMutableList()
-
-    val state = DiffState(options)
-
-    undecidedFromBaseline.each { result ->
-        val foundInReport = result.equalIndicators
-            .flatMap(reportIndex::getOrEmpty)
-            .filter(undecidedFromReport::remove)
-            .onEach { state.put(it, BaselineState.UNCHANGED) }
-            .firstOrNull() != null
-
-        when {
-            foundInReport -> remove()
-            !options.wasChecked.apply(result) -> {
-                remove()
+        .filter { result ->
+            val foundInReport = result.equalIndicators
```

Comment:
it means we are comparing hashes v1 against v2. It shouldnt be really bad for performance. But it could be important for correctness.

Reply:
Changed that. For now it doesn't affect correctness, because the hash algorithms are not compatible with each other. An alternative approach could also be to first determine which version of fingerprint to use, but I don't think it's necessary atm.